### DISCUSSION TITLE
How to manipulate data after Pydantic type validates/parses input

### DISCUSSION THREAD

**Discussion by alukach** [pydantic/pydantic#2716] [Question] [answered]
**TLDR:** I have input data like `{"country": "USA", "emissions": "[{\"date\": \"2020-01-01\", \"value\": 1.23}]"}`, what's the right way to use Pydantic to coerce that data to `{"country": "USA", "emissions": [["2020-01-01", 1.23]]}`?

---

I have a DB query that returns JSON string like `"[{\"date\": \"2020-01-01\", \"value\": 1.23}]"`. I need to parse this string as JSON and return it as part of a model for a FastAPI endpoint.  It's amazing that Pydantic supports parsing JSON with the `Json` type:

```py
from typing import List
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Json

class Emission(BaseModel):
    date: date
    value: Decimal

class CountryAgg(BaseModel):
    country: str
    emissions: Json[List[Emission]] 
```

This works! However, I don’t actually want the output to be `[{"date": "2020-01-01", "value": 1.23}]`. I want to reduce that object to be just a list of tuples, like `[["2020-01-01", 1.23]]`.  I can achieve this with a validator:

```py
from typing import List
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Json, validator

class Emission(BaseModel):
    date: date
    value: Decimal

class CountryAgg(BaseModel):
    country: str
    emissions: Json[List[Emission]]  # TODO: Docs show Emission object, not post-validated tuple

    @validator("emissions", each_item=True)
    def check_names_not_empty(cls, v: Agg) -> Tuple[date, Decimal]:
        return (v.date, round(v.value, 3))
```

This also works!  We can mutate the output with a validator. However, Pydantic/FastAPI doesn’t know about the fact that emissions is now a `Tuple[date, float]` and not a `List[Emission]`, so my auto-generated docs is wrong :scream:.

It feels like the right way to do this would be to create a custom type that inherits from `Json[List[Emission]]` and calls a method on `super()` to do all the magic that `Json[List[Emission]]` provides, but then does some other custom parsing to return `Tuple[date, Decimal]`. However, I don't know enough about Pydantic and creating custom types to achieve this.

So basically, my question is:

* What's the right way to convert data _after_ Pydantic does its magic with `Json[List[Emission]]` in a way so that something that uses my Pydantic model for documentation (e.g. FastAPI's autogenerated docs) will be aware of the final format of the data?

**Comment by PrettyWood** [comment, marked as answer]
@alukach You should put the desired inner types in your models and handle serialization with validators
Does something like this work as expected?
```py
from datetime import date
from decimal import Decimal
from typing import Any, Dict, List, NamedTuple, Union

from pydantic import BaseModel, validator


class Emission(NamedTuple):
    date: date
    value: Decimal

class CountryAgg(BaseModel):
    country: str
    emissions: List[Emission]

    @validator('emissions', each_item=True, pre=True)
    def validate_emission(cls, v: Union[Emission, Dict[str, Any]]) -> Emission:
        return Emission(**v) if isinstance(v, dict) else v


data = '{"country": "USA", "emissions": [{"date": "2020-01-01", "value": 1.23}]}'
agg = CountryAgg.parse_raw(data)
assert agg.emissions[0].date == date(2020, 1, 1)
assert agg.emissions[0].value == Decimal('1.23')
assert agg.dict() == {'country': 'USA', 'emissions': [Emission(date=date(2020, 1, 1), value=Decimal('1.23'))]}
assert agg.json() == '{"country": "USA", "emissions": [["2020-01-01", 1.23]]}'
assert CountryAgg.schema() == {
    "title": "CountryAgg",
    "type": "object",
    "properties": {
        "country": {"title": "Country", "type": "string"},
        "emissions": {
            "title": "Emissions",
            "type": "array",
            "items": {
                "type": "array",
                "items": [
                    {"title": "Date", "type": "string", "format": "date"},
                    {"title": "Value", "type": "number"}
                ]
            }
        }
    },
    "required": ["country", "emissions"]
}
```

**Reply by alukach** [reply]
Thanks for the reply @PrettyWood!  Unfortunately, this would not work, the input data is not as you presented it.  It is actually like this (if in string form):

```py
data = '{"country": "USA", "emissions": "[{\"date\": \"2020-01-01\", \"value\": 1.23}]"}'
```

This is why I needed the `Json` class to do its magic.

**Reply by PrettyWood** [reply]
Can't you change to
```py
class CountryAgg(BaseModel):
    country: str
    emissions: List[Emission]

    @validator('emissions', pre=True)
    def validate_emission(cls, v: Union[List[Emission], str]) -> List[Emission]:
        return [Emission(**c) for c in json.loads(v)] if isinstance(v, str) else v
```
?
My point is that you should focus on the inner types, not the outer type. Especially if that's what you want to export in your JSON schema. Serialization can be customized very easily thanks to `validator`

**Reply by alukach** [reply]
Thanks so much for sticking with me on this topic!  This last example really helped me get my head straight.

I think my main failing was because I was modeling the _input_ type, not what I wanted the _output_ type to be.    This may have come from thinking about the `Json` type, which seems particularly related to the _input_ data format rather than its output.

Here's what I ended up with. It mainly differs from what you suggested being that my goal is to have a `Tuple` as an output:

```py
Emission = Tuple[date, Decimal]


class CountryAgg(BaseModel):
    country: str
    emissions: List[Emission]

    @validator("emissions", pre=True)
    def parse_emissions(cls, v: Union[List[Emission], str]) -> List[Emission]:
        if not isinstance(v, str):
            return v
        return [(obj["date"], round(obj["value"], 3)) for obj in json.loads(v)]
```

I'm all ears if you have any recommendations about how to be idiomatic about my approach, but what I've got now definitely solves my problems. Thanks again for all the help!
